### PR TITLE
[Consensus] Active Pacemaker: `Forks` cleanup

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/dapperlabs/flow-go/issues/6292

### Context

This PR implements forks cleanup as described in issue. Removing unnecessary logic that we don't need anymore. 
There is a question how to proceed with event `OnQcIncorporated`, this was called by `AddQC` function which was removed, need to decide if we still need it.

### CODE REVIEW

File: consensus/hotstuff/consumer.go [hunk 1]
```diff
@@ -133,14 +133,6 @@ type Consumer interface {
 	// and must handle repetition of the same events (with some processing overhead).
 	OnQcIncorporated(*flow.QuorumCertificate)
```
Comment (durkmurder) [inline consensus/hotstuff/consumer.go:134]: Previously this one was emitted by `ForkChoice`, we don't have that anymore, we could transfer this responsibility to `Forks` to notify about every QC incorporated in `AddBlock` function. We could also leave it for later and have a separate issue to update/extend/remove all metrics.

**Review by jordanschalm — APPROVED** [review]

File: consensus/hotstuff/consumer.go (hunk 1 above)
Comment (jordanschalm) [reply consensus/hotstuff/consumer.go:134]: From what I can tell, it's not used at all currently. My gut feeling would be to remove it, if we don't have a specific use case for making use of the notifications.

Comment (AlexHentschel) [reply consensus/hotstuff/consumer.go:134]: +1

**Review by AlexHentschel — APPROVED** [review]
Looks very good. I have put up a quick PR https://github.com/onflow/flow-go/pull/2724 (targeting your branch with a few minor extensions of goDoc and one added nil-pointer check). 

On top of the changes in your PR, I would like to request a more aggressive cleanup and decluttering of `Forks` and the associated logic. I am happy for you to merge your PR and then maybe put up a follow-up PR; also happy if you include the additional cleanup here. 

## Suggestions for further cleanup 

* interfaces [`Finalizer`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer.go) and [`Forks`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go) are now largely redundant:
   * The methods `IsSafeBlock` and `LockedBlock` from `Finalizer` are not used anymore. They could be removed.
   * The methods `VerifyProposal` and `AddProposal` from `Finalizer` are commonly called right after each other. I think we could merge their logic 
* [`ForksReader`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go#L32) is never used as a stand-alone interface. I feel the methods from `ForksReader` could be moved into [`Forks`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go#L17)
* The [`Forks` implementation](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/forks.go) is largely a wrapper around `Finalizer` which doesn't add much functionality (despite combining the `Finalizer`'s method `VerifyProposal` and `AddProposal`)

I feel all this adds layers ob abstraction that obfuscate the core logic. I think this is undesired snd should be cleaned up. Suggested concrete steps:
1. Move all methods from [`ForksReader`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go#L32)  into [`Forks`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go#L17); remove `ForksReader` interface. 
2. remove [`Finalizer`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer.go) interface and replace all of its uses by [`Forks`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go#L17)
3. replace [`Forks` implementation](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/forks.go) by [`Finalizer` implementation](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/finalizer.go)
   * I would suggest to move [`finalizer.go`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/finalizer.go)  and [`blockcontainer.go`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/blockcontainer.go) into package `consensus/hotstuff/forks` (where we now have the `forks.go` implementation)
   * rename the file `finalizer.go` to `forks.go`
   * rename the `Finalizer struct` to `Forks` 
4. Declutter [`Finalizer` implementation](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/finalizer.go) (now `Forks struct`):
   * remove `IsSafeBlock`
   * remove `LockedBlock`
   * remove `LockedBlockQC`
   * remove `FinalizedBlockQC`
5. Merge methods `VerifyProposal` and `AddProposal` in [`Finalizer` implementation](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/finalizer.go) (now `Forks struct`):
   *  Rename [`AddProposal`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/finalizer.go#L137) to `UnverifiedAddProposal`.
   * Add a new method  `AddProposal` that first calls `VerifyProposal` and subsequently `UnverifiedAddProposal` (you could copy the [logic from the old Forks](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/forks.go#L49-L58))
6. At this point, there is still some legacy logic that calls `VerifyProposal` first, which we now have already integrated into `AddProposal`. We should clean this up: 
    * remove this explicit [verification step in the `FollowerLogic`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/follower/follower.go#L68-L73)
    * side note: the recovery logic in the [consensus `Follower`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/recovery/follower.go#L26) and [consensus `Participant`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/recovery/participant.go#L26) previously added the blocks without any sanity check. As we integrated the sanity check now into `AddProposal` there is a marginal runtime overhead compared to the old code. I don't think this makes any difference, because this is a one-time cost during startup. Furthermore, the `Forks` data structure lives entirely in-memory and operations on it are _very_ fast.

File: consensus/hotstuff/eventhandler/event_handler.go [hunk 2]
```diff
@@ -249,10 +249,7 @@ func (e *EventHandler) startNewView() error {
 
 		// as the leader of the current view,
 		// build the block proposal for the current view
-		qc, _, err := e.forks.MakeForkChoice(curView)
-		if err != nil {
-			return fmt.Errorf("can not make fork choice for view %v: %w", curView, err)
-		}
+		qc := e.paceMaker.NewestQC()
 
 		proposal, err := e.blockProducer.MakeBlockProposal(qc, curView)
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/eventhandler/event_handler.go:254]: correct!

File: consensus/hotstuff/eventhandler/event_handler_test.go [hunk 3]
```diff
@@ -195,23 +195,23 @@ func (v *Voter) IsSafeToTimeout(curView uint64, newestQC *flow.QuorumCertificate
 	panic("to be implemented")
 }
 
-// Forks mock allows to customize the Add QC and AddBlock function by specifying the addQC and addBlock callbacks
+// Forks mock allows to customize the Add QC and AddBlock function by specifying the addQC and addProposal callbacks
 type Forks struct {
 	mocks.Forks
 	// blocks stores all the blocks that have been added to the forks
-	blocks    map[flow.Identifier]*model.Block
+	blocks    map[flow.Identifier]*model.Proposal
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/eventhandler/event_handler_test.go:202]: I will apply those suggestions in event handler PR, afraid of creating lots of conflicts

File: consensus/hotstuff/eventhandler/event_handler_test.go [hunk 4]
```diff
@@ -237,9 +238,9 @@ func NewForks(t require.TestingT, finalized uint64) *Forks {
...
 
-func (f *Forks) AddBlock(block *model.Block) error {
-	log.Info().Msgf("forks.AddBlock received Block for view: %v, qc: %v\n", block.View, block.QC.View)
-	return f.addBlock(block)
+func (f *Forks) AddProposal(proposal *model.Proposal) error {
+	log.Info().Msgf("forks.AddProposal received Proposal for view: %v, qc: %v\n", proposal.Block.View, proposal.Block.QC.View)
+	return f.addProposal(proposal)
 }
 
 func (f *Forks) AddQC(qc *flow.QuorumCertificate) error {
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply consensus/hotstuff/eventhandler/event_handler_test.go:246]: Yup, this is all done in EventHandler PR

**Comment by durkmurder** [comment]
> Looks very good. I have put up a quick PR #2724 (targeting your branch with a few minor extensions of goDoc and one added nil-pointer check).
> 
> On top of the changes in your PR, I would like to request a more aggressive cleanup and decluttering of `Forks` and the associated logic. I am happy for you to merge your PR and then maybe put up a follow-up PR; also happy if you include the additional cleanup here.
> 
> ## Suggestions for further cleanup
> _Observations_:
> 
> * interfaces [`Finalizer`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer.go) and [`Forks`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go) are now largely redundant:
>   
>   * The methods `IsSafeBlock` and `LockedBlock` from `Finalizer` are not used anymore. They could be removed.
>   * The methods `VerifyProposal` and `AddProposal` from `Finalizer` are commonly called right after each other. I think we could merge their logic
> * [`ForksReader`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go#L32) is never used as a stand-alone interface. I feel the methods from `ForksReader` could be moved into [`Forks`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go#L17)
> * The [`Forks` implementation](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/forks.go) is largely a wrapper around `Finalizer` which doesn't add much functionality (despite combining the `Finalizer`'s method `VerifyProposal` and `AddProposal`)
> 
> I feel all this adds layers ob abstraction that obfuscate the core logic. I think this is undesired snd should be cleaned up.
> 
> _Suggested concrete steps:_
> 
> 1. Move all methods from [`ForksReader`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go#L32)  into [`Forks`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go#L17); remove `ForksReader` interface.
> 2. remove [`Finalizer`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer.go) interface and replace all of its uses by [`Forks`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks.go#L17)
> 3. replace [`Forks` implementation](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/forks.go) by [`Finalizer` implementation](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/finalizer.go)
>    
>    * I would suggest to move [`finalizer.go`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/finalizer.go)  and [`blockcontainer.go`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/blockcontainer.go) into package `consensus/hotstuff/forks` (where we now have the `forks.go` implementation)
>    * rename the file `finalizer.go` to `forks.go`
>    * rename the `Finalizer struct` to `Forks`
> 4. Declutter [`Finalizer` implementation](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/finalizer.go) (now `Forks struct`):
>    
>    * remove `IsSafeBlock`
>    * remove `LockedBlock`
>    * remove `LockedBlockQC`
>    * remove `FinalizedBlockQC`
> 5. Merge methods `VerifyProposal` and `AddProposal` in [`Finalizer` implementation](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/finalizer.go) (now `Forks struct`):
>    
>    * Rename [`AddProposal`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/finalizer/finalizer.go#L137) to `UnverifiedAddProposal`.
>    * Add a new method  `AddProposal` that first calls `VerifyProposal` and subsequently `UnverifiedAddProposal` (you could copy the [logic from the old Forks](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/forks/forks.go#L49-L58))
> 6. At this point, there is still some legacy logic that calls `VerifyProposal` first, which we now have already integrated into `AddProposal`. We should clean this up:
>    
>    * remove this explicit [verification step in the `FollowerLogic`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/hotstuff/follower/follower.go#L68-L73)
>    * side note: the recovery logic in the [consensus `Follower`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/recovery/follower.go#L26) and [consensus `Participant`](https://github.com/onflow/flow-go/blob/9714dffc8518b3ff53e600e2df133b46ef5ccfcb/consensus/recovery/participant.go#L26) previously added the blocks without any sanity check. As we integrated the sanity check now into `AddProposal` there is a marginal runtime overhead compared to the old code. I don't think this makes any difference, because this is a one-time cost during startup. Furthermore, the `Forks` data structure lives entirely in-memory and operations on it are _very_ fast.

I would like to split this cleanup in few PRs. 
Specifically I will handle everything forks related in this PR, everything finalizer related in [this](https://github.com/dapperlabs/flow-go/issues/6191) issue.

**Comment by durkmurder** [comment]
bors merge

**Pull request merged**